### PR TITLE
Docker examples

### LINKED ISSUE #27: Creating a docker image for sktime-dl
This certainly seems a prudent next step in aiding usage/setup, particularly for non-expert users 

Or any alternative/similar systems to use instead? I noticed while setting this issue up that sktime has a neat binder plugin on the readme, worth investigating for sktime-dl too @mloning?

### PR DESCRIPTION
**Pull request by Withington**
#### Reference Issues/PRs
Fixes #27 and #26.

#### What does this implement/fix? Explain your changes.
Dockerfile added. Jupyter notebook time_series_classification added. 

README.rst provides instructions for building the Docker image and then using it to either run the tests or to install the local sktime-dl project and then launch Jupyter.

#### Any other comments?
Draft pull request: I will add more text to the notebook, a regression and a multivariate classifier.

This was the best way I found to enable local sktime-dl to be installed in the Jupyter environment. Even so, it only works for notebooks that are saved at the top folder level (hence the Dockerfile ```COPY examples/time_series_classification.ipynb .```). Any better ways to do this?

### CODE REVIEW

**Comment by Withington** [comment]
Forecasting and regression examples added. Multivariate example to be added once multivariate capability is merged into dev branch.

**Comment by mloning** [comment]
FYI I'm currently refactoring the forecasting API in base sktime (see [#218](https://github.com/alan-turing-institute/sktime/pull/218)), so we may have to do some changes once I'm done with that.

**Comment by Withington** [comment]
Ok, no problem.

**Comment by James-Large** [comment]
Nice. Things are getting busy again, but task for me this week is get set up using docker and test this out

**Comment by Withington** [comment]
When merged to dev, the Binder badge in `README.rst` needs to be updated to point to the correct repo and branch i.e. to
`.. image:: https://mybinder.org/badge_logo.svg
    :target: https://mybinder.org/v2/gh/sktime/sktime-dl/dev?filepath=examples
`

The build passed on my fork.

**Comment by mloning** [comment]
Let's run Binder on the more stable master branch, but can update this in our merge from dev to master.

File: sktime_dl/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+
+FROM tensorflow/tensorflow:2.1.0-gpu-py3
+RUN pip install tensorflow-addons==0.8.2
+RUN pip install Cython==0.29.14
+RUN pip install pytest==5.3.5
+RUN pip install pytest-cov==2.8.1
+RUN pip install flaky==3.6.1
+RUN pip install sktime==0.3.0
```
Comment (mloning) [inline sktime_dl/Dockerfile]:
```suggestion
RUN pip install sktime==0.3.1
```

File: README.rst [hunk 2]
```diff
@@ -4,7 +4,8 @@
     :target: https://badge.fury.io/py/sktime-dl
 .. image:: https://badges.gitter.im/sktime/community.svg
```
[the remark this answers is not in the document]
Comment (mloning) [reply README.rst]: We can also make the pypi and travis badge nicer, take a look at the sktime readme [here](https://github.com/alan-turing-institute/sktime/blob/dev/README.rst)

**Comment by James-Large** [comment]
Nice, opened up the binder plugin and ran through the examples, everything works. 

I'm waiting for permission from my uni to have full admin rights on my machine, since too many things (docker for windows, now) require it. Have generally avoided it so far (I typically remote in when working at home), but may just set all of this up on my home PC tomorrow

Have you gotten the docker images up and running @mloning ? I don't imagine it not working either way, the whole point being platform independent, but good to test either way

With the readme, yeah I think we'll need to fine-tooth comb it just on dev once we decide for a master push, can tidy the installation instructions, we'll have a couple more networks to list, and can talk about the contributor (e.g. any links you'd want on there as well as your name @Withington ?)

**Comment by Withington** [comment]
Please note that changing the binder button to `master` breaks the button for now. To try it on binder please use:
https://mybinder.org/v2/gh/sktime/sktime-dl/dev?filepath=examples 
or 
https://mybinder.org/v2/gh/Withington/sktime-dl/docker_examples?filepath=examples

**Comment by Withington** [comment]
Just @Withington on the README is fine for me, thanks.

**Comment by mloning** [comment]
When building the docker image, I ran into the error

> ERROR: Could not find a version that satisfies the requirement sktime==0.3.1 (from versions: 0.1.dev0, 0.1.0, 0.2.0, 0.3.0)
> ERROR: No matching distribution found for sktime==0.3.1

Probably because we're not distributing the right wheels for linux yet (see [#188](https://github.com/alan-turing-institute/sktime/issues/188)). Sorry, that was my suggestion to change it. After changing it back, it worked, but haven't run the built image yet.

File: sktime_dl/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+
+FROM tensorflow/tensorflow:2.1.0-gpu-py3
+RUN pip install tensorflow-addons==0.8.2
+RUN pip install Cython==0.29.14
+RUN pip install pytest==5.3.5
+RUN pip install pytest-cov==2.8.1
+RUN pip install flaky==3.6.1
+RUN pip install sktime==0.3.1
```
Comment (mloning) [inline sktime_dl/Dockerfile]:
```suggestion
RUN pip install sktime==0.3.0
```

Comment (Withington) [reply sktime_dl/Dockerfile]: No worries.

**Comment by Withington** [comment]
This Docker branch is good to go from my point of view. Are you happy to merge it into dev?

File: sktime_dl/Dockerfile (hunk 1 above)
Comment (mloning) [inline sktime_dl/Dockerfile:8]: Could we use a common `requirements.txt` file for this? I've added one in PR #50.

Comment (Withington) [reply sktime_dl/Dockerfile:8]: Yes, no problem. I'll take a look later in the week.

**Comment by James-Large** [comment]
Yep, re-using the requirements.txt makes sense, otherwise happy for this to be merged, good job

**Comment by Withington** [comment]
Added build_tool/requirements.txt, copied from #50 

As before, launch binder with https://mybinder.org/v2/gh/Withington/sktime-dl/docker_examples?filepath=examples 

See README for instructions on running in Docker.

NB the Dockerfile is in sktime-dl/sktime_dl, not sktime-dl because otherwise Binder picks up the Dockerfile and then fails because it is not set up to create a container for Binder.

**Comment by mloning** [comment]
Looks good! 

One last thing: The docker-binder conflict should be solved by including the `.binder` folder from #50 which takes precedence, perhaps also copy that over, so that the Dockerfile still sits in the root directory, otherwise I'd suggest moving it to `build_tools` instead

**Comment by James-Large** [comment]
Merging #50 introduced a conflict of an extra blank line in requirements.txt :man_facepalming:

Assuming tests work, good to merge this now given the previous .binder commit?

**Comment by Withington** [comment]
Binder isn't working with .binder/requirements.txt . Right now the notebook launches but univariate_time_series_classification.ipynb says `ModuleNotFoundError: No module named 'sktime_dl'` . 

When it had sktime-dl in .binder/requirements.txt it couldn't find sktime_dl.deeplearning so maybe it was loading from pip. I took it out of req but now I have the error above.

Any ideas?

**Comment by mloning** [comment]
* That's because now there is no sktime-dl in the `.binder/requirements` file. 
* Binder can either build from source or we can just pip install whatever we need. Since we want to run these notebooks on stable versions, I think simply pip installing the latest sktime-dl version is the easiest, plus anything else that we need for the notebooks, e.g. matplotlib for plotting. So the binder requirements are different from the build requirements. 
* We may have to keep Cython in there for now as pip installing sktime-dl may tries to build sktime from source if no matching distribution can be found. 
* I suggest to revert to the `.binder/requirements` file I pushed in #50 and then it'll work when we push everything to master and make a new release.

**Comment by Withington** [comment]
Done, thanks @mloning . If the tests pass, it's ready to merge.

**Comment by James-Large** [comment]
Good stuff, merging now

**Comment by mloning** [comment]
@Withington I saw that the docker ignore file not in the root directory, does that matter? Or should we move it there?

**Comment by Withington** [comment]
Yes I should have moved it when I moved the Dockerfile. Let me know if you need me to move it or to test it and I’ll jump on to my dev box.

**Pull request merged**